keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -90,23 +90,52 @@ We propose to
+   This means in detail for primitive types and no types (i.e. IntOrString or arbitrary JSON) in the structural schema:
+   
+   * `if _, ok := obj[fld]; !ok` => default
+   * `if !nullable && v == nil` => default
+   
+   and for `array` type in the schema one of these:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `if !nullable && v == nil` => default
+   * `if array, ok := v.([]interface{}); !ok` => return deserialization error
+   * `if len(array) == 0` => default
+  
+   and for `object` type in the schema:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `if !nullable && v == nil` => default
+   * `if object, ok := v.(map[string]interface{}); !ok` => return deserialization error
+   * `if len(object) == 0` => default.
+   
+Hence, we never default zero numbers, integers or empty string if they are specified explicitly in the input.
 
 Note: the behaviour for empty lists and maps is that of native resources in by far the most cases. If we need to preserve empty lists and maps, we can add a vendor extensions `x-kubernetes-default-preserve-empty: true` without breaking the existing behaviour.
 
-We do this in the serializer by passing a real defaulter to [`versioningserializer.NewCodec`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49) such that defaulting is done natively just after the binary payload has been unmarshalled into an `map[string]interface{}` and pruning of [KEP: Pruning for CustomResources](https://github.com/kubernetes/enhancements/pull/709) was done, compare the yellow boxes in the following figure:
+We do this in the serializer by passing a real defaulter to [`versioningserializer.NewCodec`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49) such that defaulting is done natively just after the binary payload has been unmarshalled into an `map[string]interface{}` and pruning of [KEP: Pruning for CustomResources](/keps/sig-api-machinery/20180731-crd-pruning.md) was done.
+
+Like for native resources, we do defaulting
+
+* during request payload deserialization
+* after mutating webhook admission
+* during read from storage.
+     
+Note: like for native resources, we do not default after webhook conversions. Hence, webhook conversions must be complete in the sense that they return defaulted objects. Technically we could do defaulting, but to match native resources, we do not.
+     
+Compare the yellow boxes in the following figure:
 
 ![Decoding steps which must apply defaults](20190426-crd-defaulting-pipeline.png)
 
-We rely on the validation steps in the request pipeline to verify that the default value is of the right type.
+We rely on the validation steps in the request pipeline to verify that the default value validates value validation. We will check the types using the _structural schema_ during CRD creation and update though.
```

Comment:
is it not possible to validate the default values completely?

Reply:
no. Value validation can depend on values on higher levels and in other subtrees. There is always the chance that it is incomplete. We can do the best we can with CRD validation.

Reply:
ah, true. best effort with the local schema subtree still seems valuable